[ABI/Metadata] Re-arrange space allocation of function metadata flags

xedin opened the pull request:
This changes layout of the `TargetFunctionTypeFlags` to allocate
8 bits for flags, 8 bits for convesion and 16 bits for number of
parameters. Documentation is updated accordingly.

Resolves: rdar://problem/31408030

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ABI/Metadata] Re-arrange space allocation of function metadata flags

This changes layout of the `TargetFunctionTypeFlags` to allocate
8 bits for flags, 8 bits for convesion and 16 bits for number of
parameters. Documentation is updated accordingly.

xedin commented:
/cc @rjmccall changed everything as we've discussed in my previous PR.

xedin commented:
@swift-ci please test

xedin commented:
Some of the tests are going to fail, just want to know how many, will fix as a follow up tonight.

xedin commented:
@swift-ci please test

rjmccall commented on docs/ABI/TypeMetadata.rst:
```diff
@@ -123,18 +123,20 @@ contain the following fields:
+  flags. If the function has multiple returns, this references a `tuple metadata`_
+  record.
+- The **parameter vector** begins at **offset 3** and consists of pointers to
+  metadata records of the function's parameters.
```
I don't think we need to say anything about "offset 3" here, but we should say something like:

> The **parameter type vector** follows the result type and consists of NumParameters type metadata pointers corresponding to the types of the parameters.

rjmccall commented on docs/ABI/TypeMetadata.rst:
```diff
@@ -123,18 +123,20 @@ contain the following fields:
+- The **parameter vector** begins at **offset 3** and consists of pointers to
+  metadata records of the function's parameters.
+- The optinal **parameter flags vector** begins after the end of **parameter vector**
+  and consists of N (where N is the number of parameters) unsigned 32-bit integer
```
"optional".  And you can just say NumParameters; people will understand where that comes from.

rjmccall commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -542,17 +542,17 @@ class ExistentialTypeFlags {
+    ConventionMask    = 0x00FF0000U,
+    ConventionShift   = 16U,
     ThrowsMask        = 0x10000000U,
     ParamFlagsMask    = 0x01000000U,
```
Are you intentionally using separate bits here instead of 0x01000000 and 0x02000000?

Also, I think we should have a comment here explaining the planned approach if we run out of bits in the FunctionTypeFlags.  (We use the last bit to say that there's another bitset stored elsewhere, and we pass it to getFunctionTypeParameter at the end of the parameter-type array.)

xedin replied:
Yes, I just wanted to keep them separate, but I can change it to 0x01 and 0x02 to keep them close together if that's better?

rjmccall replied:
Please.  Using adjacent bits avoids creating unnecessary problems if we someday want a multi-bit field.

xedin replied:
Alright, no problem!

xedin commented:
@rjmccall Ok, I've updated the doc with your suggestions and changed flags to be close together + added comment about running out of space, please take a look one more time!

xedin commented:
@swift-ci please test

rjmccall commented on docs/ABI/TypeMetadata.rst:
```diff
@@ -122,19 +122,21 @@ In addition to the `common metadata layout`_ fields, function metadata records
-
-  If the function takes no arguments, **argument vector** is going to be empty.
+  includes flags - 'throws' bit, parameter flags bit (8 bits), function convention (8 bits)
+  and number of parameters (16 bits).
```
Why are you saying that the parameter flags bit is 8 bits?

xedin replied:
Let me re-phrase that a bit if that's unclear - it should be - includes flags (8 bits) which consists of 'throws' bit and 'parameter flags' bit, function convention (8 bit) etc.

rjmccall commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -542,19 +542,22 @@ class ExistentialTypeFlags {
   Swift = 0,
   Block = 1,
-  Thin = 2,
+  Thin  = 2,
```
This seems unnecessary.  It's important to line up bitmask constants, not all constants.

xedin replied:
Ok, let me remove that indent.

rjmccall reviewed (commented):
Just minor things; otherwise LGTM, thanks!

xedin commented:
@swift-ci please test and merge

xedin commented:
@swift-ci please test and merge
